### DISCUSSION TITLE
GlobalOpenTelemetry returns no-op when using Agent

### DISCUSSION THREAD

**Discussion by dreedyman** [open-telemetry/opentelemetry-java-instrumentation#12916] [Q&A]
I've created an extension where I implement a `AutoConfigurationCustomizerProvider` to setup OpenTelemetry SDK. The extension is embedded into the agent. All works as planned. However, when I attach the agent to a JVM, and from within that app I invoke `GlobalOpenTelemetry.get()`, I get a no-op instance. I had expected to obtain the configured SDK that my extension had created.

I've followed the discussion [here](https://github.com/open-telemetry/opentelemetry-java/pull/5010#issuecomment-1340390631) but must be missing something. Any suggestions welcome.

**Comment by laurit** [comment]
By any chance are you using `otel.instrumentation.common.default-enabled=false`? Firstly try running without extension does it work then? If it does then it must be something that your extension is doing.

**Comment by dreedyman** [comment]
@laurit That is indeed the case, I have set `otel.instrumentation.common.default-enabled=false` to turn off everything, then selectively turn on what instrumentations I need. After removing that system property, the `GlobalOpenTelemetry.get()` returns as expected.

Is there another option to do what I'm intending to do? I'd rather not have all instrumentations available, just those that teams need, and still be able to have `GlobalOpenTelemetry.get()` available. Thank you!

**Reply by laurit** [reply]
`GlobalOpenTelemetry.get()` is instrumented to expose the OpenTelemetry instance configured by the agent. You have disabled that instrumentation.

**Reply by dreedyman** [reply]
@laurit As documented [here](https://opentelemetry.io/docs/zero-code/java/agent/disable/#enable-only-specific-instrumentation), this property is used  to disable all default auto instrumentation, and then we can selectively re-enable individual instrumentation. It  shouldn't turn the agent off. The agent still instruments for the instrumentations that have been enabled using `otel.instrumentation.[name].enabled=true`.

Are you saying that if I want `GlobalOpenTelemetry.get()` to return as expected I have to enable _all_ instrumentations? Doesn't this contradict the advice given that control what instrumentations are enabled, to assist with the performance and reduction of possible duplicate spans?

**Reply by laurit** [reply]
> Are you saying that if I want GlobalOpenTelemetry.get() to return as expected I have to enable all instrumentations?

You have to enable only the instrumentation that instruments `GlobalOpenTelemetry.get()`. Disabling instrumentations by default and selectively enabling some isn't a great idea precisely because for average user it is hard to assess which instrumentations they need and which instrumentations depend on other instrumentations.

**Reply by dreedyman** [reply]
> You have to enable only the instrumentation that instruments `GlobalOpenTelemetry.get()`. 

Aha.  Perhaps if it were pointed out either online or in this discussion that `GlobalOpenTelemetry.get()`  is not only created using the logic explained [here](https://github.com/open-telemetry/opentelemetry-java/pull/5010#issuecomment-1340390631), but also controlled by the `otel.instrumentation.opentelemetry-api.enable` property (I guess you did,  I just didn't pick it up). Did I get that right? 

> Disabling instrumentations by default and selectively enabling some isn't a great idea precisely because for average user > it is hard to assess which instrumentations they need and which instrumentations depend on other instrumentations.

Well, this somewhat contradicts the documented approach on suppressing and enabling the instrumentations that have been identified as being required.

And if that's all correct, then we can close this, and thank you for the assistance.

**Reply by trask** [reply]
see (and vote!) for https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/7201
